### PR TITLE
Add Consmax and script for testing

### PR DESCRIPTION
**Pull request by gkielian**
This is Shiwei's ConSmax, a Softmax alternative with learnable parameters for beta ('xmax') and gamma (denominator sum of terms).

Script included will run this and produce tensorboard logs to show perplexity comparisons on the default dataset.

### CODE REVIEW

**Comment by gkielian** [comment]
Note, I will try to rebase to remove the csv forecasting files from the pr commit history

**Comment by gkielian** [comment]
Added the latest test script and running currently around 30 times to get a good idea.

Also adding a temperature hyperparameter.

I still have to add the prints for the hyperparams over time, should add this this weekend.

**Comment by gkielian** [comment]
Latest updates include looping over softmax variants, further parameterizing the script, and adding a means to initialize the gamma values directly in the model.py Constantmax nn.Module.

Note: a means for exposing these to the train.py have not yet been added.

**Comment by gkielian** [comment]
Seems the merge conflicts are increasing, I'll probably try to base the changes here onto the main repo branch to prevent the merge conflicts, and may create a new PR if it comes to it.

**Comment by gkielian** [comment]
Moving to https://github.com/ReaLLMASIC/nanoGPT/pull/50

This has the core of this PR, I think its better to have a separate PR for post_ln option and addition of the random_seed training option.
